README.md
# NOMADe-Data-Visualization
MATLAB Scripts for Visualization of Measurement Data




changelog.md
# Changelog NOMADe data visualization
All notable changes to the [NOMADe data visualization](https://github.com/DimitriDeSchuyter/NOMADe-Data-Visualization.git) are listed.

<br/>
## [1.0.3] - 2022-08-04
Added the local database option 

## [1.0.2] - 2022-04-08
Added the following data output datatypes:
-  186 	IMU_GYRO_ACC_MAG (0xBA)
- 	187 	IMU_GYRO_ACC_MAG_100Hz (0xBB)

## [1.0.1] - 2021-11-25
Added the following data output datatypes:
-  181 	IMU_QUAT_GYRO_ACC (0xB5)
- 	182 	IMU_QUAT_GYRO_ACC_100Hz (0xB6)
- 	183 	IMU_QUAT_100Hz (Quat only) (0xB7)
- 	184 	IMU_QUAT_9DOF (Quat only) (0xB8)
- 	185 	IMU_QUAT_9DOF_100Hz (Quat only) (0xB9)

## [1.0.0] - 2021-07-02
 
**Initial commit**


+classes/data.m
classdef data
    % Teste ne
    
    properties
        name, ...
        values, ...
        datatype, ...
        unit
    end
    methods
        function obj = data(name, unit, datatype,maxCycleCount)
            obj.name = name;  
            obj.unit = unit;
            switch datatype %% ENUMERATING to prevent String compares during live loop
                case "boolean"
                    obj.datatype = 0;
                    obj.values = NaN(maxCycleCount,1);
                case "int_8"
                    obj.datatype = 1;
                    obj.values = NaN(maxCycleCount,1);
                case "uint_8"
                    obj.datatype = 2;  
                    obj.values = NaN(maxCycleCount,1);
                case "int_16"
                    obj.datatype = 3;
                    obj.values = NaN(maxCycleCount,1);
                case "uint_16"
                    obj.datatype = 4;
                    obj.values = NaN(maxCycleCount,1);
                case "iint_32"
                    obj.datatype = 5;
                    obj.values = NaN(maxCycleCount,1);
                case "uint_32"
                    obj.datatype = 6;
                    obj.values = NaN(maxCycleCount,1);
                case "float_32"
                    obj.datatype = 7;
                    obj.values = NaN(maxCycleCount,1);
                case "int_64"
                    obj.datatype = 8;
                    obj.values = NaN(maxCycleCount,1);
                case "uint_64"
                    obj.datatype = 9;
                    obj.values = NaN(maxCycleCount,1);
                case "float_64"
                    obj.datatype = 10;
                    obj.values = NaN(maxCycleCount,1);
                otherwise
                    error("MATLAB Datatype not supported")
            end
        end
        
        function obj = add_value(obj,cyclecounter_list, blob)
            switch obj.datatype %% ENUMERATING to prevent String compares during live loop
                case 0 %boolean/already converted
                    obj.values(cyclecounter_list) = blob;
                case 1 %int_8
                    reOut = reshape(blob',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'int8'); 
                case 2 %uint_8
                    reOut = reshape(blob',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'uint8');
                case 3 %int_16
                    reOut = reshape(blob(:,2:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'int16');
                case 4 %uint_16
                    reOut = reshape(blob(:,2:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'uint16');
                case 5 %int_32
                    reOut = reshape(blob(:,4:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'int32');
                case 6 %uint_32
                    reOut = reshape(blob(:,4:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'uint32');
                case 7 %float_32
                    reOut = reshape(blob(:,4:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'single');
                case 8 %int_64
                    reOut = reshape(blob(:,8:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'int64');
                case 9 %uint_64
                    reOut = reshape(blob(:,8:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'uint64');
                case 10 %float_64
                    reOut = reshape(blob(:,8:-1:1)',1,[]);
                    obj.values(cyclecounter_list) = typecast(reOut,'double');
                    %obj.values = typecast(blob(8:-1:1),'double');
            end
            %obj.value(cyclecounter).value;
            
            
            
            
        end
        
        function obj = remove_value(obj,cyclecounter_list)

            obj.values= obj.values(cyclecounter_list);
            
        end
        function obj = filteredData(obj,cyclecounter,data)
            obj.values(cyclecounter) = data;
        end
        
        function obj = resize(obj)
            %obj.value = obj.value;
        end
        function obj = plot(obj,startTime,enableTitle,enableXlabel,yEnLim,plotDownSample,downSampleFactor)
            time = seconds(0:(size(obj.values,1)-1))*0.020 + startTime;
             if plotDownSample
                factor = size(obj.values,1)/downSampleFactor;
                plt(time,obj.values,'LineWidth',2,'downsample',factor);
            else
                plot(time,obj.values,'LineWidth',2);
             end
            if enableTitle
                
                    subtitle([char(obj.name) ' [' char(obj.unit) ']'],'fontsize',20);
                                
            end
            if yEnLim && obj.datatype ~= 0
                ylim([min([-0.1; obj.values],[],'all')*1.1 (1.1*max([obj.values;0.1],[],'all'))]);
            elseif obj.datatype == 0
                ylim([-0.1 1.1]);
            end
            xlim(([min(time) (max(time))]));
            if enableXlabel
                tzOffset = tzoffset(startTime);
                xlabel('Time (UTC +'+  extractBefore(string(tzOffset,'hh:mm'),':')+ ')','fontsize',20);
            end
            if strlength(char(obj.name))> 30
                yText = split(extractAfter(char(obj.name),strlength(char(obj.name))/2),['-',' ']);
                yText = char(yText(end));
            else
                yText = char(obj.name);
            end
            ylabel([yText ' [' char(obj.unit) ']'],'fontsize',20);
        end
    end
end

functions/popup.m
function popup(m,fctn)
    fig = uifigure;
    
     % Desired Window width and height
      GuiWidth = 500;
      if contains(fctn,'plot')
          instrumentList = isprop(m.instruments,'extracted');
          n_instruments = sum(instrumentList);
          instrumentList= find(instrumentList==1);
          lblText = 'Indicate which instruments that must be plotted:';
      else
          instrumentList = 1:m.n_instruments;
          n_instruments = m.n_instruments;
          lblText = 'Indicate which instruments that should not be retrieved:';
      end
      if isprop(m,'OAS')
        n_instruments = n_instruments +1;
    end
      GuiHeight = 30+50+n_instruments*30+10;

    % Find Screen Resolution
      temp = figposition([0,0,100,100]);
      ScreenWidth = temp(3);
      ScreenHeight = temp(4);
  
    set (fig, 'position', [ScreenWidth/2 - GuiWidth/2, ScreenHeight/2 - GuiHeight/2, GuiWidth, GuiHeight]);

    fig.Name = 'NOMADe extraction option';
    height= (n_instruments+1)*30+20;
    lbl = uilabel(fig);
    lbl.Text = lblText;
    lbl.FontSize = 14;
    lbl.FontWeight= 'bold';
    lbl.Position = [10 height 400 30];
    cb = gobjects(n_instruments,1);
    for ii = instrumentList
      cb(ii) = uicheckbox('Parent',fig,'Position',[20 height - ii*30 200 30]);
      cb(ii).Text = m.instruments(ii).name; 
    end
    if isprop(m,'OAS')
        cb(ii+1) = uicheckbox('Parent',fig,'Position',[20 height - (ii+1)*30 200 30]);

      cb(ii+1).Text = 'OAS'; 
    end

    % Select all button
    btn = uibutton('Parent',fig,'Position',[10 15 80 30],'Text','Select all',...
        'ButtonPushedFcn', @(btn,event) SelectAll(btn, cb,instrumentList,m));

    % Confirm button
    uibutton('Parent',fig,'Position',[GuiWidth-90 15 80 30],'Text','Confirm','FontWeight','bold',...
                'ButtonPushedFcn', @(btn,event) plotButtonPushed(cb,instrumentList,m));
            uiwait (fig);   
end



% Create the function for the ButtonPushedFcn callback
function plotButtonPushed(cb,instrumentList,m)
    includedInstruments= zeros(1,size(instrumentList,2));
    for i=instrumentList
        includedInstruments(i) = cb(i).Value;
    end
    if isprop(m,'OAS')
     includedInstruments(i+1) = cb(i+1).Value;
    end
     assignin('base','includedInstruments',includedInstruments);
     closereq;
end

function SelectAll(btn,cb,instrumentList,m)
    if btn.Text == "Select all" 
        state = 1;
         btn.Text = "Deselect all";
    else
        state = 0;
        btn.Text = "Select all";
    end
    for ii = instrumentList
          cb(ii).Value =state;
    end
    if isprop(m,'OAS')
        cb(ii+1).Value =state;
    end
    
end


functions/extractSubmissions.m
% ------------------------------------------------------------------------------------
%---------------  Extraction of NOMADe submissions  ---------------
% ------------------------------------------------------------------------------------



% create filename
filename =[datestr(now,'yyyy-mm-dd HH-MM')   '_questionnaires.xlsx'];
 filename = fullfile(pwd,filename);
 i=0;
  while exist(filename, 'file') == 2
      i = i+1;
      filename =[datestr(now,'yyyy-mm-dd HH-MM')   '_questionnaires V' int2str(i) '.xlsx'];
        filename = fullfile(pwd,filename);
  end
  
          %% Connection to the database
  
            if isfile("../jdbc/mysql-connector-java-8.0.18.jar")
                javaclasspath("../jdbc/mysql-connector-java-8.0.18.jar");
            elseif isfile("../mysql-connector-java-8.0.18/mysql-connector-java-8.0.18.jar")
                % legacy jdbc location
                javaclasspath("../mysql-connector-java-8.0.18/mysql-connector-java-8.0.18.jar");
            else
                error("JDBC MySQL Connector not found, please download the connector from https://dev.mysql.com/downloads/connector/j/ and extract it in the 'jdbc' directory. Note: the mysql-connector-java-8.0.18.jar musn't be placed in a subdirectory, but directly in the root of the jdbc directory.");
            end

            databaseName = "nomade";
            username = "nomade-dev";
            disp([' NOMADe DB username: ' char(username)]);
            if isfile("password.mat")
                load password.mat password;
            end
            if ~exist('password','var')
                password = input(' NOMADe DB password: ','s');
                store_password = input(' store this password (Y/N): ','s');
                if store_password == "Y" || store_password == "y" || store_password == "yes"
                    save password.mat password;
                    warning off backtrace;
                    warning("Password saved to 'password.mat' file");
                    warning on backtrace;
                end
            else
                disp(' NOMADe DB password (retrieved from file)');
            end
            
            jdbcDriver = "com.mysql.cj.jdbc.Driver";
            server = "jdbc:mysql://clouddb.myriade.be:20100/";
            
            
            conn = database(databaseName, username, password, jdbcDriver, server);
            %% Get submissions
         startSubmissionID= 600;
         endSubmissionID= 623;
         
                    sqlquery = ['SELECT TRIM(`submissions`.`id`) AS `submissionID`, '...
                     '               TRIM(`users`.`username`) AS `username`, '...  
                     '               TRIM(`t4`.`name_en`) AS `questionnaire`, '...  
                     '               TRIM(`t`.`question_id`) AS `questionID`, '... 
                     '               TRIM(`t2`.`name_en`) AS `question`, '...        
                     '               TRIM(`t3`.`name_en`) AS `prefix answer`, '...    
                     '              TRIM(`t`.`value`) AS `answer text`, '...    
                     '              TRIM(`submissions`.`created_at`) AS `submission date` '...    
                     '                 FROM `submissions` '...
                     ' INNER JOIN `submission_answers`  AS `t` '...
                     ' ON `submissions`.`id` = `t`.`submission_id` '...
                     ' INNER JOIN `questions`  AS `t2` '...
                     ' ON `t`.`question_id` = `t2`.`id`  '...
                     ' INNER JOIN `answers`  AS `t3` '...
                     ' ON `t`.`answer_id` = `t3`.`id` '...     
                     ' INNER JOIN `users`   '...
                     ' ON `submissions`.`user_id` = `users`.`id` '...
                     ' INNER JOIN `questionnaires` AS `t4`   '...
                     ' ON `submissions`.`questionnaire_id` = `t4`.`id` '...
                     '          WHERE (`submissions`.`id` BETWEEN ' int2str(startSubmissionID) ' AND ' int2str(endSubmissionID) ');'];  
                     
                   extract = select( conn,sqlquery);
             
         

   
        writetable(extract,filename,'Range','A1')


functions/getAllMeasurements.m
import classes.*
addpath( 'functions')
m = measurement();

m = m.connect('ID','ASC');
getList =m.list.id;
for id = 20:numel(getList)
    disp( "******* Start measurement " +getList(id) + " ********")
    m = measurement();
    m = m.connect('ID','ASC');
    m=m.set_measurement_ID(getList(id));
    disp( "******* Declaration ********")
    m = m.declaration("full",0,true);
     disp( "******* Extract and process data ********")
     includedInstruments = zeros(1,m.n_instruments);
    m = m.get_dataset_DB(includedInstruments,true);
    disp( "******* Saving to .mat ********")
    mextracted = m.extractionMobject();
    storeName = strcat('ID',num2str(m.id),'_m-extracted_',datestr(m.start_time,'yyyy_mm_dd_HHMMSS'));
     store = fullfile('H:\5. NOMADe\backups\measurementObjects', [storeName '.mat']);
    warning off backtrace;
    save(store,'mextracted','-v7.3');
    warning on backtrace;
    clearvars -except getList
end

functions/measurementCutting_popup.m
function executed = measurementCutting_popup(m)
fig = uifigure;
 fig.CloseRequestFcn = @(fig,event)my_closereq(fig,false);
% Desired Window width and height
GuiWidth = 400;
GuiHeight = 250;

% Find Screen Resolution
temp =get(0,'screensize');
ScreenWidth = temp(3);
ScreenHeight = temp(4);

set (fig, 'position', [ScreenWidth/2 - GuiWidth/2, ScreenHeight/2 - GuiHeight/2, GuiWidth, GuiHeight]);

fig.Name = 'NOMADe measurement option';

lbl = uilabel(fig);
lbl.Text = "Choose start and end time";
lbl.FontSize = 14;
lbl.FontWeight= 'bold';
lbl.Position = [110 220 400 30];

lblStrt = uilabel(fig);
lblStrt.Text = "New start time";
lblStrt.FontSize = 14;
lblStrt.Position = [10 148 100 30];
lblStrtTime = uilabel(fig);
lblStrtTime.Text = "Current start time: " + datestr(m.start_time,'dd/mm/yyyy HH:MM:SS.FFF');
lblStrtTime.FontSize = 14;
lblStrtTime.Position = [10 180 320 20];

strtTime = uieditfield(fig,...
    'Position',[110 150 280 22],...
    'ValueChangedFcn',@(strtTime,event) startTimeChanged(strtTime,m));
strtTime.Value = datestr(m.start_time,'dd/mm/yyyy HH:MM:SS.FFF');

lblEnd = uilabel(fig);
lblEnd.Text = "New end time";
lblEnd.FontSize = 14;
lblEnd.Position = [10 78 100 30];
lblEndTime = uilabel(fig);
lblEndTime.Text = "Current end time: " + datestr(m.end_time,'dd/mm/yyyy HH:MM:SS.FFF');
lblEndTime.FontSize = 14;
lblEndTime.Position = [10 110 320 20];
endTime = uieditfield(fig,...
    'Position',[110 80 280 22],...
    'ValueChangedFcn',@(endTime,event) endTimeChanged(endTime,m));

endTime.Value = datestr(m.end_time,'dd/mm/yyyy HH:MM:SS.FFF');
% plot button
uibutton('Parent',fig,'Position',[(GuiWidth/2-100) 30 80 30],'Text','plot','FontWeight','bold',...
    'ButtonPushedFcn', @(btn,event) plotSignals(m));
% Confirm button
uibutton('Parent',fig,'Position',[(GuiWidth/2+20) 30 80 30],'Text','Confirm','FontWeight','bold',...
    'ButtonPushedFcn', @(btn,event) confirmButtonPushed(fig,m,strtTime,endTime));
uiwait (fig);
end



% Create the function for the ButtonPushedFcn callback
function confirmButtonPushed(fig,m,strtTime,endTime)
            startTimeConv= convertTime(strtTime.Value);
            endTimeConv= convertTime(endTime.Value);
            % start time of measurement to the next sample if start
            % time is not a multiple of 0.02
            startDiff =time2num(startTimeConv  - m.start_time,"seconds");
             endDiff =time2num(m.end_time  - endTimeConv,"seconds");
             ret = false;
            if startDiff <-3.552713678800501e-18
                strtTime.BackgroundColor = 'r';
                ret = true;
            end
             if endDiff <-3.552713678800501e-18
                endTime.BackgroundColor = 'r';
                ret = true;
             end
            if ret 
                return
            end
            
assignin('base','startTime', strtTime.Value);
assignin('base','endTime', endTime.Value);

my_closereq(fig,true);

end


% Create ValueChangedFcn callback
function startTimeChanged(timeField,m)
    startTimeConv= convertTime(timeField.Value);
     startDiff =time2num(startTimeConv  - m.start_time,"seconds");
        if startDiff <0
            timeField.BackgroundColor = 'r';
            return
        else
            timeField.BackgroundColor = 'w';
        end
end

function endTimeChanged(timeField,m)
    endTimeConv= convertTime(timeField.Value);
     endDiff =time2num(m.end_time-endTimeConv,"seconds");
        if endDiff <0
            timeField.BackgroundColor = 'r';
            return
        else
            timeField.BackgroundColor = 'w';
        end
end


function plotSignals(m)
figure;
fontSize= 20;
set(gcf,'Visible','on') % show outside live script
set(gca,'fontsize',fontSize+2) % set fontsize of the plot to 20
set(gcf,'units','normalized','outerposition',[0 0 1 1]) % full screen
set(0, 'DefaultAxesFontSize', fontSize);
subplotArray(1) = subplot(2,1,1);
m.instruments(1).data(2).plot(m.start_time,true,false,true,false,0);
subplotArray(2) = subplot(2,1,2);
m.instruments(1).data(3).plot(m.start_time,true,false,true,false,0);
linkaxes(subplotArray,'x');
end
function state = my_closereq(fig,st)
assignin('base','cuttingState', st);
delete(fig)
end



function convTime =  convertTime(time)
% split date and reformat the date
            dateSplit = ["00","00","0000","00","00","00","000"];
            splitted = split(time,{'-','/',':','h','m','s','.',' '},2);
            dateSplit(1:size(splitted,2)) = splitted;
            dateSymbols = ["-","-"," ",":",":",".",""];
            dateArray = reshape([dateSplit(1:7) ;dateSymbols(1:7)],1,[]);
            dateConverted = strjoin(dateArray(1:end-1),"");   
            convTime=datetime(dateConverted,'InputFormat','dd-MM-yyyy HH:mm:ss.SSS','TimeZone','local');
end